### ISSUE TITLE
Creation of too big multidimensional array returns empty tensor.

### ISSUE THREAD

**Issue by andrewnaguib** [pytorch/pytorch#20056]
## 🐛 Bug

When trying to create a tensor of too many dimensions it simply returns an empty tensor with its shape is the dimensions I passed even though it should have contained the value `1`.

## To Reproduce

Number of 2's in the following example is `>63`.
```Python
torch.ones((2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2),
dtype=torch.uint8)
```
Output: 

```Python
tensor([],
       size=(2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2),
       dtype=torch.uint8)
```

Another related issue

```Python
torch.ones((2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2),
dtype=torch.uint8)
```

These are `62` 2's. 

Output: 

```Python
RuntimeError: $ Torch: not enough memory: you tried to allocate 0GB. Buy new RAM! at /opt/conda/conda-bld/pytorch_1549628766161/work/aten/src/TH/THGeneral.cpp:201
```

When they're `63` 2's. The output changes to:

```Python
RuntimeError: $ Torch: invalid memory size -- maybe an overflow? at /opt/conda/conda-bld/pytorch_1549628766161/work/aten/src/TH/THGeneral.cpp:188
```

## Expected behavior

I expected to see the aforementioned dimensions filled with `1` if it's possible, or an exception that says a value error.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```
```
PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Arch Linux
GCC version: (crosstool-NG 1.23.0.449-a04d0) 7.3.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration: GPU 0: GeForce GTX 1050 Ti with Max-Q Design
Nvidia driver version: 418.56
cuDNN version: /usr/lib/libcudnn.so.7.5.0

Versions of relevant libraries:
[pip3] numpy==1.16.2
[pip3] torch==1.0.1.post2
[pip3] torchvision==0.2.2.post3
[conda] _tflow_select             2.3.0                       mkl  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.1                      144  
[conda] mkl_fft                   1.0.10           py36ha843d7b_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] pytorch                   1.0.1           py3.6_cuda9.0.176_cudnn7.4.2_2    pytorch
[conda] tensorflow                1.12.0          mkl_py36h69b6ba0_0  
[conda] tensorflow-base           1.12.0          mkl_py36h3c3e929_0  
[conda] torchvision               0.2.1                      py_2    pytorch
```
## Additional context

I suspect that the desired behavior is not happening because tensors seem to be basing on NumPy `ndarray` which accepts only `32` dims.

Also, I have just found this [line](https://github.com/pytorch/pytorch/blob/7ddd5d06ed07a50b94aa6b2fdffa2f667d677c4b/aten/src/TH/THGeneral.cpp#L184).

```cpp
THError("$ Torch: not enough memory: you tried to reallocate %dGB. Buy new RAM!", size/1073741824);
```

I think the integer division of those two elements results in the error.

**Comment by andrewnaguib** [comment]
@ifedan I'm not sure how you guys triage things. But, I wanted to tell you that I edited the issue in case you want to edit labels.

**Comment by andrewnaguib** [comment]
@soumith I'm not sure If I should be mentioning you, but I just wanted to know the updates on this. I did not receive any comments on that.

**Comment by soumith** [comment]
I dont think it's been resolved yet. I dont think anyone is actively working on it at the moment either

**Comment by soumith** [comment]
cc: @gchanan

**Comment by gchanan** [comment]
I was able to confirm the first issue on master, if you would like to look into fixing it, @andrewnaguib we would be happy to give you guidance.

I wasn't able to reproduce the second issue, it looks to have been fixed:
```
      1 torch.ones((2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2),
----> 2 dtype=torch.uint8)

RuntimeError: [enforce fail at CPUAllocator.cpp:64] . DefaultCPUAllocator: can't allocate memory: you tried to allocate 4611686018427387904 bytes. Error code 12 (Cannot allocate memory)
```

The third issue still seems relevant, though:
```
In [5]: torch.ones((2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2
   ...: , 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2),
   ...: dtype=torch.uint8)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-796b36549665> in <module>()
      1 torch.ones((2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2),
----> 2 dtype=torch.uint8)

RuntimeError: [enforce fail at CPUAllocator.cpp:47] ((ptrdiff_t)nbytes) >= 0. alloc_cpu() seems to have been called with negative number: 9223372036854775808
```

**Comment by andrewnaguib** [comment]
Yes, I would. Thanks!  I will look into the 1st and 3rd issues, and will ensure I have the latest version (anyways) and recheck the 2nd. Where do you suggest I start looking?

**Comment by fmassa** [comment]
I don't think this would be something easy fixing. IIRC  TensorIterator has assumptions on the maximum number of dimensions that a Tensor can have.

**Comment by andrewnaguib** [comment]
@fmassa

**Comment by andrewnaguib** [comment]
@fmassa Are you saying that the first issue has a direct connection with changing the maximum number of dimensions? I only thought the intention would be raising a `RuntimeError`. Also, where the maximum number of dims defined?

**Comment by gchanan** [comment]
raising a `RuntimeError` sounds good.  Silently giving you an empty tensor is bad.

**Comment by andrewnaguib** [comment]
@gchanan Yeah. However, does that mean a tensor has a maximum number of dims?

**Comment by gchanan** [comment]
I'm not sure if we generally enforce that at construction time.  Do you get the same behavior if you use `torch.empty` ?

**Comment by fmassa** [comment]
Here is an example where we enforce a max number of dimensions
https://github.com/pytorch/pytorch/blob/eabfca3577ea85df2d68bdf747c62dd4a5fff5cf/aten/src/ATen/cuda/detail/OffsetCalculator.cuh#L15
This affect most point wise operations in PyTorch

**Comment by fmassa** [comment]
But I think I misunderstood the goal. If raising an error at creation time is the goal, then I'm ok with it. But making PyTorch support tensors with arbitrary number of dimensions is more complicated. It used to be supported in the past, but we dropped support for it last year

**Comment by andrewnaguib** [comment]
@gchanan Yes, same behavior exist with `torch.empty`. Also, confirmed the 1st and 3rd issues on the master (at the time of writing this reply).  

I believe the both issues happens because of the unexpected behavior of casting:

https://github.com/pytorch/pytorch/blob/4976208e739a05a7cf6309f33f3b07ad0cd8e50e/c10/core/CPUAllocator.cpp#L48

Because of an overflow; since if you try `std::numeric_limits<std::ptrdiff_t>::max()`, the result will be `9223372036854775807` (x64) which is less than `9223372036854775808`. 

Also, if you try to case any number greater than >2<sup>64</sup>, `ptrdiff_t` will simply return `0` which is, indeed, `True` in the `>=0` part. (`=`) is unnecessary anyways because it is caught in:

https://github.com/pytorch/pytorch/blob/4976208e739a05a7cf6309f33f3b07ad0cd8e50e/c10/core/CPUAllocator.cpp#L42-L44

Can you confirm, please?

Should I add unit tests for these issues? 

---

@fmassa How does the `MAX_DIMS` you mentioned differ from the one in:

https://github.com/pytorch/pytorch/blob/e44b2b72bd4ccecf9c2f6c18d09c11eff446b5a3/torch/csrc/utils/tensor_new.cpp#L46

Although it is more reasonable to raise an error message at the current situation, what limits supporting larger number of dims (are the release notes for dropping support exist for checking?)? Also, why it is not explicitly stated in the docs?

**Comment by andrewnaguib** [comment]
@gchanan can you confirm, please?

**Comment by andrewnaguib** [comment]
@soumith Would it be possible to have someone just to confirm on the issue resolution in my previous comment so that I can prepare a PR?

**Comment by gchanan** [comment]
@ngimel is looking into it.

**Comment by ngimel** [comment]
There are several issues here
1) MAX_DIMS=128 in tensor_new.cpp is definitely wrong, pytorch doesn't support 128-dim tensors anywhere else, as @fmassa pointed out the number of dimensions on cuda e.g. is limited to 25. We should change MAX_DIMS in tensor_new.cpp. 
2) @andrewnaguib is correct, there's no reason casting nbytes to ptrdiff_t for comparison https://github.com/pytorch/pytorch/blob/4976208e739a05a7cf6309f33f3b07ad0cd8e50e/c10/core/CPUAllocator.cpp#L48. `posix_memalign` expects `size_t` argument. On a 64-bit system max value for size_t is 2**64-1, and max_value for ptrdiff_t is 2**63-1. Avoiding this cast will make the check correctly fire when number of elements is between 2**63 and 2**64. Comparison to 0 is also not necessary, as it was taken care of before
3) That said, we still cannot guarantee that `nbytes` computation will be correct because it can overflow in size_t, and we didn't find a palatable way to address it, see #42582 and linked PR.

**Comment by ndrwnaguib** [comment]
@ngimel I see that `MAX_DIMS` has been modified already on `master`. I will prepare a PR that removes the casting of `nbytes` into `ptrdiff_t` and the unnecessary check for equivalence to `0`, i.e., in `>= 0` to `> 0`. 

if `nbytes` overflows in `size_t`, wouldn't the result be `0` (on GCC 11.1.0) and will be caught in `if (nbytes == 0)`?
